AnisotropyBarnLamp sample model

echadwick-wayfair opened the pull request:
A sample model for the extension KHR_materials_anisotropy. Details in the readme.

![screenshot_Large](https://user-images.githubusercontent.com/40177036/208192389-da32b734-e55a-44ca-8f19-45bc66147a02.jpg)

pixnblox commented:
Excellent information in the README.md... thank you for providing that! 💯

zeux commented:
`anisotropyRotation` in this model is a single-element array, which contradicts the specification (which requires a scalar)

zeux commented:
I also should note that the contents of the anisotropy RG channels seems a little suspicious. I would expect the values on the edges of the disk to go between 0 on one side of the disk to 255 on the other side, but they look as if they are in 0..64 range. I could be mistaken here, as this is based on skimming the spec and looking at the pixels in GIMP, not actual rigorous visualization.

echadwick-wayfair commented:
OK I checked the values of a random pixel, and [the fixed texture](https://github.com/DrX3D/glTF-Sample-Assets/pull/12) seems to be correct.

![AnisotropyBarnLamp_anisotropy png](https://github.com/KhronosGroup/glTF-Sample-Models/assets/40177036/c7ef0eb3-7776-4512-a1da-128568bf0d0a)

x2 + y2 = n2
So the XY normalized values will be x/n and y/n.

1. Convert RGB values into X and Y values:
a. 218 / 256 = 0.8515625
b. 130 / 256 = 0.5078125
2. Square the X value:
a. 0.8515625 * 0.8515625 = 0.72515869140625
3. Square the Y value:
a. 0.5078125 * 0.5078125 = 0.25787353515625
4. Add the two quared values to get the N magnitude:
a. 0.72515869140625 + 0.25787353515625 = 0.9830322265625
5. Divide the squared X value by the N magnitude to get the normalized X value:
a. 0.72515869140625 / 0.9830322265625 = 0.7376754004718738
6. Divide the squared Y value by the N magnitude to get the normalized Y value:
a. 0.25787353515625 / 0.9830322265625 = 0.2623245995281262
7. Add the two normalized values to check if the normal is normalized:
a. 0.7376754004718738 + 0.2623245995281262 = 1

zeux commented:
> The spec used to specify a scalar, but was changed to use a number instead. https://github.com/KhronosGroup/glTF/tree/main/extensions/2.0/Khronos/KHR_materials_anisotropy

Sorry, maybe I wasn't clear here, but here's the snippet from the attached model (either from this PR or from DrX3D PR):

```
                "KHR_materials_anisotropy": {
                    "anisotropyStrength": 0.75,
                    "anisotropyRotation": [
                        0.785398
                    ],
                    "anisotropyTexture": {
                        "index": 3
                    }
                },
```

Here's the snippet from the spec:

```
{
    "materials": [
        {
            "extensions": {
                "KHR_materials_anisotropy": {
                    "anisotropyStrength": 0.6,
                    "anisotropyRotation": 1.57,
                    "anisotropyTexture": {
                        "index": 0
                    }
                }
            }
        }
    ]
}
```

Note that the file in this PR incorrectly specifies `anisotropyRotation` as an array - it should be a plain number.

zeux commented:
As for normalization process above, you need to check the magnitude of the vector after converting to [-1, 1] from [0, 1]. So for the specified pixel:

1. Convert RGB values into X and Y values (0..1 range):
a. 218 / 255 = 0.854901961
b. 130 / 255 = 0.509803922

2. Convert X/Y values from 0..1 range to -1..1 range as per spec:
a. 218 / 255 * 2 - 1 = 0.709803922
b. 130 / 255 * 2 - 1 = 0.019607844

3. Compute X^2 + Y^2 = 0.504206075. Expected value is around 1.

elalish commented:
I'd also recommend @donmccurdy's lossless png compression (cuts the size by half): https://github.com/mrdoob/three.js/pull/26058#issuecomment-1549940115: 
`npx gltf-transform png AnisotropyBarnLamp.glb AnisotropyBarnLampOptimized.glb`

echadwick-wayfair commented:
OK thanks for the reviews. 
1. I'll look into renormalizing to -1,1 which should be do-able in Substance Designer (normalize as before, then multiply by 2 and subtract 1). 
2. I'll remove the anisotropyRotation since it's not necessary here. 
3. I'll compress the PNGs.
4. I'll look into increasing tesslation for the base.
5. There's a tangent seam on the bottom of the flange connecting the arm to the base. I'll try re-UVing that section to make it contiguous instead.

Hopefully the glTF Sample VIewer will have anisotropy support soon, so I can test before committing changes.

The model looks really dark in these areas. Maybe non-aniso reflections are being disabled?
![anisotropy_dark_threejs](https://github.com/KhronosGroup/glTF-Sample-Models/assets/40177036/226bb8c3-16b4-4b23-956e-a9ae2fe4035e)

elalish commented:
Since I merged anisotropy support into three.js, you can load the pre-release version here: https://raw.githack.com/mrdoob/three.js/dev/editor/index.html - just set `Environment` to `MODELVIEWER` so you get some lighting.

Anisotropy changes the reflections, it doesn't add to them, so yes, the non-aniso reflections are disabled in a sense (correctly). I noticed those dark patches too - is nothing visible in the base color or AO textures? They look very desaturated - I suppose it *could* be a weird Fresnel thing...

elalish commented:
Maybe you mean the clearcoat layer? That still seems to be working. The base color map and ORM map definitely have different values in those dark areas, but the desaturation still seems really extreme. If you happen to have time to bake up a simpler repro model that would also be awesome.

echadwick-wayfair commented:
> Anisotropy changes the reflections, it doesn't add to them, so yes, the non-aniso reflections are disabled in a sense (correctly). 

What values should be used in the anisotropyTexture if we want some parts of the model to have non-anisotropic reflections? Maybe (.5,.5,1)?

elalish commented:
> What values should be used in the anisotropyTexture if we want some parts of the model to have non-anisotropic reflections? Maybe (.5,.5,1)?

`(0.5, 0.5, 0)` - no strength means it'll behave like the extension isn't present on that part.

elalish commented:
Hmm, actually maybe `(1, 0, 0)` would be better - that's a normalized direction at least. It's possible I might be getting a NaN from dividing by zero if the direction is dead center.

zeux commented:
One downside with (1, 0, 0) is that filtering will warp the direction around the zones without anisotropic reflections - not a problem if the UV shells are distinct, but if there's a fade-in zone it might look odd. Unsure how significant of a problem this is in practice, and shader normalization resulting in NaNs is also a very valid concern.

echadwick-wayfair commented:
Changes since last time:
1. There was some bad triangulation on the base, this has been fixed. 
2. The arm flange near the base had a big UV seam, now fixed.
3. The anisotropy texture has been recreated and re-normalized.

I ran a test of compressing PNGs with gltf-transform. It looks like this is creating artifacts for the anisotropy though, so I won't commit the compressed version.

![2023-05-30 15_59_26-Window](https://github.com/KhronosGroup/glTF-Sample-Models/assets/40177036/892772be-16b7-4931-9837-b2319fa792bc)
gltf-transform compressed PNGs on the left, Photoshop/Substance3DDesigner compressed PNGs on the right.

```
Microsoft Windows [Version 10.0.19044.2965]
(c) Microsoft Corporation. All rights reserved.

C:\Temp>npx @gltf-transform/cli@latest png AnisotropyBarnLamp.glb AnisotropyBarnLampOptimized.glb
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.
Need to install the following packages:
  @gltf-transform/cli@3.4.0
Ok to proceed? (y) y
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
info: AnisotropyBarnLamp.glb (8.43 MB) → AnisotropyBarnLampOptimized.glb (4.39 MB)
npm notice
npm notice New minor version of npm available! 9.2.0 -> 9.6.7
npm notice Changelog: https://github.com/npm/cli/releases/tag/v9.6.7
npm notice Run npm install -g npm@9.6.7 to update!
npm notice

C:\Temp\>npm install -g npm@9.6.7
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.

removed 5 packages, and changed 81 packages in 5s

27 packages are looking for funding
  run `npm fund` for details

C:\Temp\>
```

echadwick-wayfair commented:
![2023-05-30 16_29_48-glTF Sample Viewer](https://github.com/KhronosGroup/glTF-Sample-Models/assets/40177036/d0cc6bd7-5e6c-4adb-9e43-87467b0d6efb)

Same asset in both Three.js and glTF Sample Viewer. 

The IBL is Microsoft's high-contrast studio lighting https://github.com/BabylonJS/Assets/blob/master/environments/studioSoftbox2Umbrellas_panorama_radiance.hdr

User cameras were roughly adjusted to match,

1. Why are the anisotropic reflections at different angles? Maybe the Three.js angles of the Environment and Background don't match?
2. Parts of the model with black in the anisotropy blue channel are appearing like dielctrics in Three.js
3. The flanged part of the base is showing triangulation artifacts in the glTF Sample Viewer.
4. Three.js is rendering the metal with higher contrast/saturation than expected. The clearcoat layer appears to be correct though.

The new GLB: 
[AnisotropyBarnLamp.zip](https://github.com/KhronosGroup/glTF-Sample-Models/files/11605357/AnisotropyBarnLamp.zip)

elalish commented on 2.0/AnisotropyBarnLamp/glTF-KTX2-Basisu/AnisotropyBarnLamp.gltf:
```diff
@@ -0,0 +1,351 @@
+            "extensions": {
+                "KHR_materials_anisotropy": {
+                    "anisotropyStrength": 0.75,
+                    "anisotropyRotation": 0.785398,
```
@echadwick-wayfair The rotation is the reason for the weird anisotropy angles - if you remove that it looks fine.

echadwick-wayfair replied:
![2023-05-31 09_54_56-glTF Sample Viewer](https://github.com/KhronosGroup/glTF-Sample-Models/assets/40177036/5b1e7b62-25f4-414a-89cc-4b680b0dae15)

![2023-05-31 09_56_07-AnisotropyBarnLamp gltf - Visual Studio Code](https://github.com/KhronosGroup/glTF-Sample-Models/assets/40177036/17d23d6a-971e-4081-aa29-9d733cb1ab1e)

Curious how the anisotropic reflections line up in glTF Sample Viewer vs. Three.js ... which is correct?

echadwick-wayfair replied:
Looks like tone mapping is enabled in Three.js, so I chose the ACES tonemapper in glTF Sample Viewer that seems closest.

![2023-05-31 10_30_13-glTF Sample Viewer](https://github.com/KhronosGroup/glTF-Sample-Models/assets/40177036/520b7d54-d90b-4da2-a400-905e6986c474)

elalish replied:
Hmm, yeah the sample viewer looks right here - I'll look into that when I get a chance. Also, the rotation is not an array - not sure if the `[0]` is confusing something.

donmccurdy commented:
> I ran a test of compressing PNGs with gltf-transform. It looks like this is creating artifacts for the anisotropy though, so I won't commit the compressed version.

Interesting... I'd intended for the default PNG compression to be lossless, but misunderstood the compression library's options, and it wasn't. Tracking in https://github.com/donmccurdy/glTF-Transform/issues/969.

Switching to lossless PNG compression, there's not much benefit on this model. WebP's lossless mode would help more (8.4 MB → 5.7 MB, no visible change) but we probably don't want to use WebP for the base sample here I'd assume.

echadwick-wayfair commented:
I've committed the updated asset to `glTF-Sample-Assets` instead of here in `glTF-Sample-Models` since the latter repo is being deprecated in favor of the former.

bghgary commented:
@echadwick-wayfair There appears to be seam here.

<img width="227" alt="image" src="https://github.com/KhronosGroup/glTF-Sample-Models/assets/19158740/6120e132-9cba-4dee-9319-5607e954adfc">

echadwick-wayfair commented:
I've revised the anisotropy texture to avoid the normalization artifacts we were seeing, updated the Substance graph to create results similar to what @donmccurdy gets with glTF.Report, and added some details to the readme to explain these choices. 

https://github.com/KhronosGroup/glTF-Sample-Assets/pull/3

![anisotropy_NormalizedColors](https://github.com/KhronosGroup/glTF-Sample-Models/assets/40177036/2511a693-c056-4b01-9589-abb850e3e101)

Thanks everyone for your help improving this asset!
